src/Microsoft.IdentityModel.Tokens/MaxCapacityRefreshCache.cs

```diff
@@ -0,0 +1,137 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// A cache that clears itself upon reaching max capacity.
+    /// </summary>
+    /// <typeparam name="TKey">The type of cache key.</typeparam>
+    /// <typeparam name="TValue">The type of cached value.</typeparam>
+    internal class MaxCapacityRefreshCache
+        <TKey, TValue>
+        where TValue : class
+    {
+        /// <summary>
+        /// Gets or sets the cache.
+        /// </summary>
+        private ConcurrentDictionary<TKey, TValue> Cache { get; set; }
+
+        /// <summary>
+        /// Gets or sets the cache size limit.
+        /// </summary>
+        private long SizeLimit { get; set; }
+
+        /// <summary>
+        /// Gets or sets the lock for recycle operations.
+        /// </summary>
+        private object RecycleLock { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MaxCapacityRefreshCache{TKey,TValue}" /> class.
+        /// </summary>
+        /// <param name="sizeLimit">The max number of items the cache is capable of holding.</param>
+        /// <param name="comparer">The string comparer.</param>
+        internal MaxCapacityRefreshCache(
+            long sizeLimit,
+            IEqualityComparer<TKey> comparer = null)
+        {
+            RecycleLock = new object();
+            SizeLimit = sizeLimit;
+            Cache = new ConcurrentDictionary<TKey, TValue>(comparer ?? EqualityComparer<TKey>.Default);
+        }
+
+        internal bool Contains(TKey key)
+        {
+            return Cache.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// Gets the value corresponding to <paramref name="value"/> from the cache.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="value">The cached value.</param>
+        /// <remarks>This method is non-blocking and thread-safe.</remarks>
+        internal bool TryGetValue(TKey cacheKey, out TValue value)
+        {
+            RecycleCacheIfNeeded();
+            return Cache.TryGetValue(cacheKey, out value);           
+        }
+
+        /// <summary>
+        /// Sets the cache value.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="value">The value.</param>
+        /// <returns><paramref name="value"/> if it was successfully set, null otherwise.</returns>
+        /// <remarks>This method is non-blocking and thread-safe.</remarks>
```

Comment:
misleading to call these non-blocking when some of the methods in ConcurrentDictionary can take a lock and block

Reply:
I removed this misleading comment from all the relevant methods.